File: PreferencesSample/app/src/main/java/com/example/preferencessample/MainActivity.java
package com.example.preferencessample;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    String str = "Initial";
    Context mContext;
    SharedPreferences sharedPreferences;
    AppDatabase db;
    ContactDao contactDao;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mContext = this;
        sharedPreferences = getSharedPreferences("my_preferences", Context.MODE_PRIVATE);
        db = MyApplication.getInstance().getDatabase();
        contactDao = db.contactDao();
//        str = sharedPreferences.getString("my_string", "Initial");
//        Toast.makeText(mContext, str, Toast.LENGTH_LONG).show();

//        if (savedInstanceState != null) {
//            str = savedInstanceState.getString("my_string");
//            Toast.makeText(this, str, Toast.LENGTH_LONG).show();
//        }

        Button button = findViewById(R.id.save_button);
        final EditText idEdit = findViewById(R.id.id_edit);
        final EditText nameEdit = findViewById(R.id.name);
        final EditText phoneEdit = findViewById(R.id.phone);
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String nameStr = nameEdit.getText().toString();
                String phoneStr = phoneEdit.getText().toString();

                Contact contact = new Contact();
                contact.name = nameStr;
                contact.phone = phoneStr;
                contactDao.insert(contact);
//                SharedPreferences.Editor editor = sharedPreferences.edit();
//                editor.putString("my_string", str);
//                editor.apply();
            }
        });


        Button remove_button = findViewById(R.id.get_contacts_button);
        remove_button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                List<Contact> contacts = contactDao.getContacts();
                for (Contact contact : contacts) {
                    Log.e("Contact", contact.id + " " + contact.name + " " + contact.phone);
                }
//                SharedPreferences.Editor editor = sharedPreferences.edit();
//                editor.remove("my_string");
//                editor.apply();
            }
        });


        Button updateButton = findViewById(R.id.update_button);
        updateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int id = Integer.parseInt(idEdit.getText().toString());
                Contact contact = contactDao.getContactById(id);
                contact.name = nameEdit.getText().toString();
                contact.phone = phoneEdit.getText().toString();
                contactDao.update(contact);
            }
        });


        Button deleteButton = findViewById(R.id.delete_button);
        deleteButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                int id = Integer.parseInt(idEdit.getText().toString());
                Contact contact = contactDao.getContactById(id);
                contactDao.delete(contact);
            }
        });
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.e("MainActivity", "onDestroy");
    }

    @Override
    protected void onRestoreInstanceState(Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        Log.e("MainActivity", "onRestoreInstanceState");
//        str = savedInstanceState.getString("my_string");
//        Toast.makeText(this, str, Toast.LENGTH_LONG).show();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString("my_string", str);
    }
}

File: KotlinSampleApplication/app/src/main/java/com/example/kotlinsampleapplication/MainActivity.kt
package com.example.kotlinsampleapplication

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.Toast
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
//    , View.OnClickListener {

    private var counter: Int = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
//        Log.e("MainActivity", "onCreate")
//        counter = savedInstanceState?.getDouble("counter") ?: Math.random() * 10


//        button1.setOnClickListener { view ->
//            counter++
//            text1.text = counter.toString()
//        }

//        button1.setOnClickListener(this)
//        text1.setOnClickListener(this)
    }


    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        Log.e("MainActivity", "onSaveInstanceState")
//        outState.putDouble("counter", counter)
    }

//    fun onClick(p0: View?) {
//        when (p0?.id) {
//            button1.id -> {
//                counter++
//                text1.text = counter.toString()
//            }
//            text1.id -> {
//                counter = 0
//                text1.text = counter.toString()
//            }
//        }
//    }

    fun onClickButton(view: View?) {
        counter++
        text1.text = counter.toString()
    }

    fun onClickText(view: View?) {
        counter = 0
        text1.text = counter.toString()
    }


//    override fun onClick(p0: View?) {
//        when (p0?.id) {
//            button1.id -> {
//                counter++
//                text1.text = counter.toString()
//            }
//            text1.id -> {
//                counter = 0
//                text1.text = counter.toString()
//            }
//        }
//    }
//    override fun onStart() {
//        super.onStart()
//        Log.e("MainActivity", "onStart")
//    }
//
//    override fun onResume() {
//        super.onResume()
//        Log.e("MainActivity", "onResume")
//    }
//
//    override fun onRestart() {
//        super.onRestart()
//        Log.e("MainActivity", "onRestart")
//    }
//
//
//    override fun onPause() {
//        super.onPause()
//        Log.e("MainActivity", "onPause")
//    }
//
//    override fun onStop() {
//        super.onStop()
//        Log.e("MainActivity", "onStop")
//    }
//
//    override fun onDestroy() {
////        Toast.makeText(this, "onDestroy", Toast.LENGTH_LONG).show()
//        Log.e("MainActivity", "onDestroy")
//        super.onDestroy()
//    }
}
